### PR TITLE
Build-over-build New Feature Proposal

### PR DESCRIPTION
**Pull request by aditi-rajawat**
Build-over-build is a new feature proposed to eliminate Jenkin's builds with bad code coverage as compared to the last successful build. It allows user to configure delta threshold values for each type of coverage, which are compared to the delta coverage of the current build. Delta coverage of a build is defined as the absolute difference between coverage of last successful build and current build. If the coverage drops more than the user-configured thresholds then the build fails.
[Build-over-build New Feature Proposal.pdf](https://github.com/jenkinsci/jacoco-plugin/files/781639/Build-over-build.New.Feature.Proposal.pdf)

### CODE REVIEW

**Comment by aditi-rajawat** [comment]
Hi..my pull request failed checks due to merge conflicts. I suppose only people with write access to the repository can resolve the conflicts. Any estimate when it will be done? Thanks.

**Comment by centic9** [comment]
The conflicts mean that you developed the changes on an older version of the project. You can and should resolve the conflicts yourself by merging the latest changes from origin/master into your feature-branch and then pushing the merge-commit to Github, the PR will be updated with those changes automatically then.

**Comment by aditi-rajawat** [comment]
Ok. Thank you for clarifying. I will resolve the conflicts soon.

**Comment by aditi-rajawat** [comment]
Please review the PR. Thank you.

**Comment by aditi-rajawat** [comment]
@centic9 Hi. Can I get an estimated date for this PR merge and jacoco plugin next release? My team plans to use the next release for our project and we are seeking an estimation. Please let me know. Thank you.

**Comment by aditi-rajawat** [comment]
Hi. I changed everything to BigDecimal for more accurate calculations of delta coverage. I agree with your second comment and have removed my name from the source code. Thanks.

**Comment by aditi-rajawat** [comment]
Hi. Can this PR be merged with the master branch if all the changes have been verified satisfactorily? I would also like to know about the next version release date. Please let me know. Thanks.

**Comment by centic9** [comment]
Sorry for taking my time here, but I cannot rush here both because I want to really understand the reason for all the changes and also because I am only working on it in my free time, which is quite limited nowadays.

Do you have an actual example where the precision of float was not enough? As we mostly operate in 0-100% with up to 2 decimal digits I don't immediately see the need here. 

What about using double? All the BigIntegers will introduce quite some more complexity and some memory overhead, I only want to introduce it if it is really necessary.

Sorry for being picky, but this plugin is used by a lot of people and I don't want to introduce more complexity/overhead/memory than necessary without a good reason.

**Comment by mheinzerling** [comment]
I'm of course a friend of BigDecimals and promote them where ever possible. But I also don't get the reason why we might need them here?
Do we need the PowerMock? I never heard of it (but I also prefer custom mocks), is this the default in Jenkins plugins or should we use EasyMock or Mockito?

File: src/main/java/hudson/plugins/jacoco/portlet/bean/JacocoDeltaCoverageResultSummary.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    // Used to extract coverage result of current and last successful build and encapsulate delta coverage values
+    public static JacocoDeltaCoverageResultSummary build(Run<?,?> run){
+        Job<?, ?> parent = run.getParent();
+        Run<?,?> lastSuccessfulBuild = parent!=null ? parent.getLastSuccessfulBuild():null;
+        JacocoCoverageResultSummary lastBuildCoverage = lastSuccessfulBuild!=null ? JacocoLoadData.getResult(lastSuccessfulBuild):new JacocoCoverageResultSummary();
+        JacocoCoverageResultSummary currentBuildCoverage = JacocoLoadData.getResult(run);
+
+        JacocoDeltaCoverageResultSummary jacocoDeltaCoverageResultSummary = new JacocoDeltaCoverageResultSummary();
+        jacocoDeltaCoverageResultSummary.instructionCoverage = lastBuildCoverage.getInstructionCoverage().subtract(currentBuildCoverage.getInstructionCoverage()).abs();
```
Comment (centic9) [inline src/main/java/hudson/plugins/jacoco/portlet/bean/JacocoDeltaCoverageResultSummary.java]: This looks strange to me, why don't we keep the negative values so we know the actual difference?

**Comment by centic9** [comment]
With commits 2b0665edd0fea7bad0f39ef3220c6edff7d13883, 95983aedbefc814021692fd6baf4bf809c31ae5c and de63db9cac9274d939c9f33fac3b41e18ff97bd2 I have now applied/merged the changes, but reverted the BigDecimal changes for now. Please test the results as it was a non-primitive merge due to other concurrent changes in the meantime. 

I also adjust the handling of abs() somewhat, I hope it still works fine.

Let me know as soon as you have verified the changes so we can prepare an official release.

**Comment by aditi-rajawat** [comment]
@centic9 My apologies for not been able to respond. My other tasks took higher priority. Thank you for reverting BigDecimal changes. I understand it was not a clever choice considering the performance issues it might cause. 

I am verifying the code from the master branch.

**Comment by centic9** [comment]
no worries, I just released this as part of 2.2.0, please try that version directly via Jenkins update if possible so we get some more testing done early on and can release fixes quickly if necessary.

**Comment by aditi-rajawat** [comment]
Okay. That's great. I will test the released version then. I just have one question. can we also increase the scale of coverage percentages displayed on Jenkins job page? If yes, I can submit the change. It is important in case of millions of lines of code.

**Comment by aditi-rajawat** [comment]
I have tested the changes using 2.2.0 release version of the plugin from jenkins. It is working as expected. Also checked the source code. Thank you for including this feature and releasing it. 

It would be good if we display coverage % with increased scale (approx 6 digits after decimal). Can I add that change?

**Comment by centic9** [comment]
Yes, sure, please create a new PR

**Comment by dbyron0** [comment]
Probably breaking the rules by posting here, but the mailing list looks dormant and I don't see a way to report an issue.  I enabled the "Record JaCoCo coverage report" post-build task in my jenkins job and it worked great.  I then clicked the "Fail the build if coverage degrades more than the delta thresholds" checkbox and ran the job again and got the following stack trace:
```
[JaCoCo plugin] Overall coverage: class: 97, method: 87, line: 93, branch: 62, instruction: 90
ERROR: Step ‘Record JaCoCo coverage report’ aborted due to exception: 
java.lang.ClassNotFoundException: org.joda.time.ReadablePartial
	at java.lang.ClassLoader.findClass(ClassLoader.java:530)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at hudson.util.MaskingClassLoader.loadClass(MaskingClassLoader.java:75)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.apache.tools.ant.AntClassLoader.findBaseClass(AntClassLoader.java:1387)
	at org.apache.tools.ant.AntClassLoader.loadClass(AntClassLoader.java:1080)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
Caused: java.lang.NoClassDefFoundError: org/joda/time/ReadablePartial
	at hudson.plugins.jacoco.portlet.bean.JacocoDeltaCoverageResultSummary.build(JacocoDeltaCoverageResultSummary.java:37)
	at hudson.plugins.jacoco.JacocoPublisher.checkBuildOverBuildResult(JacocoPublisher.java:703)
	at hudson.plugins.jacoco.JacocoPublisher.perform(JacocoPublisher.java:637)
	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:78)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:779)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:720)
	at hudson.model.Build$BuildExecution.post2(Build.java:186)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:665)
	at hudson.model.Run.execute(Run.java:1753)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
```
I'm using version 2.46.1 of jenkins and version 2.2.0 of the jacoco plugin.

Thanks for your help.

**Comment by dbyron0** [comment]
And...happily now there's https://github.com/jenkinsci/jacoco-plugin/pull/87 and https://issues.jenkins-ci.org/browse/JENKINS-43103.  Thank you.

**Comment by dcalap** [comment]
Hi, I have a question/problem with this feature. I have this config in my pipeline:
` jacoco buildOverBuild: true,  deltaMethodCoverage: '1'`
And when I run a build log shows:
```
[JaCoCo plugin] Delta coverage: class: 0.0, method: -5.0, line: -1.989708, branch: 0.0, instruction: -1.413044, complexity: -4.545452
[JaCoCo plugin] Delta thresholds: JacocoHealthReportDeltaThresholds [deltaInstruction=0.0, deltaBranch=0.0, deltaComplexity=0.0, deltaLine=0.0, deltaMethod=1.0, deltaClass=0.0]
[JaCoCo plugin] Results of delta thresholds check: FAILURE
```
But build not failing and finish successfully. Any idea how to break the build if delta thresholds check FAILS?

Thanks!

**Comment by sivajangam** [comment]
screen shot 2017-12-07 at 5 09 15 pm
HI,

  I downloaded the plugin and shows different label,please find in my screenshot.
Is this the one with feature Build over Build ? if it is then what are the metrics which I need to define to make the build fail when code coverage degrades compared to last successful build coverage

**Comment by centic9** [comment]
Yes, the "fail the build if " is probably what you are looking for.

**Comment by sivajangam** [comment]
But what should be the ideal metrics which I need to give, as 2.0 for all or any other metrics, as 0 might fail

**Comment by sivajangam** [comment]
Can anyone explain me the scenario, as I kept delta threshold as 2.0,
as per formula D coverage = | coverage of last successful build – coverage of current build |

But for me the build was success in scenario-1 and failing in scenario-2.

**Scenario-1**

02:16:01 [JaCoCo plugin] Delta coverage: class: 1.900177, method: 2.404953, line: 1.9257431, branch: 0.84700394, instruction: 0.9057579, complexity: 1.597332
02:16:01 [JaCoCo plugin] Delta thresholds: JacocoHealthReportDeltaThresholds [deltaInstruction=2.0, deltaBranch=2.0, deltaComplexity=2.0, deltaLine=2.0, deltaMethod=2.0, deltaClass=2.0]
02:16:01 [JaCoCo plugin] Results of delta thresholds check: SUCCESS

**Scenario -2**

11:40:54 [JaCoCo plugin] Delta coverage: class: -1.900177, method: -2.404953, line: -1.9257431, branch: -0.84700394, instruction: -0.9057579, complexity: -1.597332
11:40:54 [JaCoCo plugin] Delta thresholds: JacocoHealthReportDeltaThresholds [deltaInstruction=2.0, deltaBranch=2.0, deltaComplexity=2.0, deltaLine=2.0, deltaMethod=2.0, deltaClass=2.0]
11:40:54 [JaCoCo plugin] Results of delta thresholds check: FAILURE

**Pull request merged**